Add dataplane performance automated test

ciarams87 opened the pull request:
### Proposed changes

Problem: Currently, all our NFR tests need to be run manually. This is time consuming every release.

Solution: Use the framework to automate one test to start off.

Testing: Ran the tests against a local kind cluster, locally against a GKE cluster, and on a GCP VM against a GKE cluster

Closes #1286 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [ ] I have rebased my branch onto main
- [ ] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

sjberman commented on tests/framework/script.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+func AddEntryToHostsFile(domain string, ip string) error {
+	hostsFile, err := os.OpenFile("/etc/hosts", os.O_APPEND|os.O_WRONLY, 0o644)
```
I wonder our approach to these tests....because editing this file as part of automation feels too privileged.

Can we just utilize port forwarding? We'd be bypassing the LB, but we're still ultimately testing the traffic flow of our product.

ciarams87 replied:
I am still utilising port forwarding when it's local, the issue I was getting though is that wrk wasn't connecting when using the IP directly instead of the hostname and it doesn't have any `--resolve` or similar option. Also, in the output, we don't see the hostname, just the IP:port. I can look again today. 

Having said that, I think the VM tests can still take this approach, as it's an ephemeral environment. These `wrk` tests don't make sense to run locally anyway, so maybe they can be skipped in those cases.

sjberman replied:
Can we avoid using scripts and `wrk` and instead just use pure Go?

pleshakov replied:
> Can we just utilize port forwarding? We'd be bypassing the LB, but we're still ultimately testing the traffic flow of our product.

port-forwarding will put a proxy between wrk and data plane, which will affect performance numbers

> Can we avoid using scripts and wrk and instead just use pure Go? Having automation edit the /etc/hosts file, even if in an ephemeral environment, feels like a security issue waiting to happen.

what if we run it in a container? so we start wrk in a container with modified /etc/hosts

sjberman replied:
> port-forwarding will put a proxy between wrk and data plane, which will affect performance numbers

If we go through the LB it's still the same issue though

pleshakov replied:
afaik, port forwarding works through Kubernetes API server through one of its go processes. So it is a constrained user land proxy.

kate-osborn replied:
> Can we avoid using scripts and wrk and instead just use pure Go?

This might work: https://github.com/form3tech-oss/f1

I like the idea of using Go

ciarams87 replied:
I looked a bit today and came across https://github.com/tsenart/vegeta - looked at f1 too, but I liked Vegeta because it provides a mechanism to provide a proxy URL which solves the /etc/hosts problem (plus it provides histogram output, ability to do distributed testing etc, which I'm not using now, but could be a nice addition later). Its output is a bit different than wrk so it's a bit of a departure - I've pushed up those results as well in the last commit.

pleshakov commented on tests/utils/run-tests-gcp-vm.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
I wonder if shell scripts and ssh'ing will be manageable as we increase the number of supported tests. 

I wonder if there are any tools we can use to:
- declaratively provision VMs and configure it with the needed config
- declaratively run a test on a VM or a set of VMs

ciarams87 replied:
Sure, there are plenty of tools 😁 We could use an IaC tool (terraform, pulumi etc) but is that too much overhead for running locally? It means more dev dependencies.

ciarams87 commented on tests/suite/results/dp-perf/ciara/ciara.md:
```diff
@@ -0,0 +1,103 @@
+# Results
```
Moved

sjberman commented on tests/Makefile:
```diff
@@ -24,8 +26,16 @@ load-images: ## Load NGF and NGINX images on configured kind cluster
+
+.PHONY: run-tests-on-vm
+run-tests-on-vm: ## Run the tests on a GCP VM
+	bash utils/run-tests-gcp-vm.sh
```
Maybe just personal preference, but I like a `scripts` directory for these instead of `utils`.

sjberman commented on tests/framework/load.go:
```diff
@@ -0,0 +1,45 @@
+	targeter := vegeta.NewStaticTargeter(targets...)
+	proxyURL, err := url.Parse(proxy)
+	if err != nil {
+		return fmt.Errorf("Error getting proxy URL: %w", err)
```
lint: errors should start with lowercase (same for all instances like this)

sjberman commented on tests/framework/load.go:
```diff
@@ -0,0 +1,45 @@
+	reporter := vegeta.NewTextReporter(&metrics)
+
+	err = reporter.Report(outFile)
+	if err != nil {
```
nit: inline error

sjberman commented on tests/framework/request.go:
```diff
@@ -11,9 +11,9 @@ import (
+// It resolves to the specified address instead of using DNS.
 // The status and body of the response is returned, or an error.
-func Get(url string, timeout time.Duration) (int, string, error) {
+func Get(url string, address string, timeout time.Duration) (int, string, error) {
```
```suggestion
func Get(url, address string, timeout time.Duration) (int, string, error) {
```

sjberman commented on tests/suite/dataplane_perf_test.go:
```diff
@@ -0,0 +1,148 @@
+	pwd, err := os.Getwd()
+	Expect(err).ToNot(HaveOccurred())
+	parentDir := filepath.Dir(pwd)
+	var port string
+	var resultsFile string
+	var addr string
```
```suggestion
	var port, resultsFile, addr string
```

sjberman commented on tests/suite/dataplane_perf_test.go:
```diff
@@ -0,0 +1,148 @@
+	})
+})
+
+func runDPPTestOutputToFile(outputPath string, ci framework.ClusterInfo, addr string) error {
```
```suggestion
func runDPPTestOutputToFile(ci framework.ClusterInfo, addr, outputPath string) error {
```

sjberman commented on tests/utils/run-tests-gcp-vm.sh:
```diff
@@ -0,0 +1,23 @@
+    --tags=${TAGS} --create-disk=auto-delete=yes,boot=yes,device-name=${VM_NAME},image=${IMAGE},mode=rw,size=10 --no-shielded-secure-boot --shielded-vtpm --shielded-integrity-monitoring --labels=goog-ec-src=vm_add-gcloud --reservation-affinity=any
+
+# Give the VM a chance to come up before attempting ssh connection
+sleep 15
```
Is there a polling way we can do this?

sjberman commented on tests/README.md:
````diff
@@ -74,10 +78,26 @@ make build-images load-images TAG=$(whoami)
+Before running the below `make` command, populate the required env vars in `utils/vars.env`.
+
+```makefile
+make run-tests-on-vm
````
If I create my GKE cluster and run this script, is there anything else I need to do? Do I need to supply an ssh key, set up firewall policy, etc?

ciarams87 replied:
I think it might depend on your configuration with GCP, but I don't need to provide any SSH key, it uses my cloud credentials I believe. The firewall policy is configured using tags - depending on what region you are deploying into, there may already be a rule you can use, otherwise you will need to create a new one. The rule name should be your TAG entry in the vars.env file (mine is `ciara-ssh` for example, but it's restrictive and only allows my IP in my region). You also need to provide a service account with the correct access - there is one available, I can provide the details offline.

kate-osborn replied:
Can the script create a firewall policy that allows ssh access from our IP only?  We did something like this before for mesh where we used `$(curl -sS -4 icanhazip.com)/32` to get the IP address of the machine running the command. It would be nice if we didn't have to create anything manually.

ciarams87 replied:
Done!

sjberman reviewed (commented):
I'd like to be able to fully remove the `dataplane-performance` directory. We can migrate the past results into the new results folder.

ciarams87 commented:
@sjberman 

> I'd like to be able to fully remove the dataplane-performance directory. We can migrate the past results into the new results folder.

I removed it in the first commit, I think you can possibly just see the files being renamed?

pleshakov commented on tests/suite/dataplane_perf_test.go:
```diff
@@ -0,0 +1,148 @@
+		},
+	}
+
+	for i, test := range tests {
```
would it make sense to run them independently?  separate Its of Ginkgo?

Gomega can also support table driven approach -- https://onsi.github.io/ginkgo/#table-specs

ciarams87 replied:
I really like that table driven ability! Changed.

pleshakov reviewed (commented):
I ran it on GCP. worked as expected

ciarams87 commented on tests/README.md:
````diff
@@ -30,23 +30,27 @@ make
 
 ```text
````
Added something, let me know if you think it's sufficient!

ciarams87 commented:
@pleshakov 

> would it simply things if we run the test in a container in a separate cluster? as a Job
>
> for local run, we can run it in the same kind cluster?
> this way, we can declaratively build infra? if we need to separate jobs to send traffic - example, one HTTP and the other HTTPS, we can create two different jobs, scheduling them on different dedicated nodes to prevent interference.
> 
> this way we will define all test deps in a Dockerfile.
> also, we will declaratively define the job to be performed. This way, we can also re-start or re-deploy it, if we misconfigured something.

I can look into that, but because we are writing files etc it's a bit complex. For now, I did split up the make targets so that the VM creation can be split from running the tests, so that they can be reran. I also added a target for adding ginkgo flags to separate test runs - we can add labels to tests specs (I added "performance" as a start) and run tests with a specific label, or we can focus, skip etc - and these could be run on different VMs easily. I'll look into how it could be improved more.

kate-osborn commented on tests/README.md:
````diff
@@ -112,8 +151,34 @@ XIt("runs some test", func(){
+or to run all tests with the label "performance":
+
+```makefile
+make test TAG=$(whoami) GINKGO_FLAGS='-ginkgo.label-filter "performance"'
````
What do you think about adding a makefile command for this? Given the different test requirements, I think we will need to run each test separately, and this might get a little annoying to copy/paste or type out.

ciarams87 replied:
I added target for GINKGO_LABEL as well, which can be set to just the label name, e.g. `make test TAG=$(whoami) GINKGO_LABEL=performance`

kate-osborn commented on tests/README.md:
```diff
@@ -74,10 +89,34 @@ make build-images load-images TAG=$(whoami)
 
+### 3b - Run the tests on a GKE cluster from a GCP VM
+
+This step only applies if you would like to run the tests from a GCP based VM. The VM should be created in the same
```
Do we need to create a VM before running the script or is the VM created by the script?

kate-osborn commented on tests/framework/resourcemanager.go:
```diff
@@ -311,3 +323,86 @@ func (rm *ResourceManager) waitForRoutesToBeReady(ctx context.Context, namespace
+// GetLBIPAddress gets the IP or Hostname from the Loadbalancer service.
+func (rm *ResourceManager) GetLBIPAddress(namespace string) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), rm.TimeoutConfig.CreateTimeout)
+	defer cancel()
```
nit: a newline after defer statements help with readability
```suggestion
	defer cancel()
	
```

kate-osborn commented on tests/framework/resourcemanager.go:
```diff
@@ -311,3 +323,86 @@ func (rm *ResourceManager) waitForRoutesToBeReady(ctx context.Context, namespace
+		if svc.Spec.Type == core.ServiceTypeLoadBalancer {
+			nsName = types.NamespacedName{Namespace: svc.GetNamespace(), Name: svc.GetName()}
+			if err := rm.waitForLBStatusToBeReady(ctx, nsName); err != nil {
+				return "", fmt.Errorf("Error getting status from LoadBalancer service: %w", err)
```
```suggestion
				return "", fmt.Errorf("error getting status from LoadBalancer service: %w", err)
```

kate-osborn commented on tests/framework/resourcemanager.go:
```diff
@@ -311,3 +323,86 @@ func (rm *ResourceManager) waitForRoutesToBeReady(ctx context.Context, namespace
+		var lbService core.Service
+
+		if err := rm.K8sClient.Get(ctx, nsName, &lbService); err != nil {
+			return "", fmt.Errorf("Error getting LoadBalancer service: %w", err)
```
```suggestion
			return "", fmt.Errorf("error getting LoadBalancer service: %w", err)
```

kate-osborn commented on tests/framework/resourcemanager.go:
```diff
@@ -311,3 +323,86 @@ func (rm *ResourceManager) waitForRoutesToBeReady(ctx context.Context, namespace
+// GetClusterInfo retrieves node info and Kubernetes version from the cluster
+func (rm *ResourceManager) GetClusterInfo() (ClusterInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), rm.TimeoutConfig.GetTimeout)
+	defer cancel()
```
```suggestion
	defer cancel()
	
```

kate-osborn commented on tests/scripts/vars.env:
```diff
@@ -0,0 +1,12 @@
+VM_NAME=ngf-auto-testing
```
I'm worried that someone might accidentally push this file filled out. Can it live outside the repo in a specified location, like `~/.ngf/vars.env ` or something?

ciarams87 replied:
I've renamed the example to `vars.env-example` and added all dotenv files to gitignore

kate-osborn commented on tests/suite/dataplane_perf_test.go:
```diff
@@ -0,0 +1,109 @@
+	targetURL := "http://cafe.example.com"
+	var outFile *os.File
+
+	t1 := vegeta.Target{
```
It would be nice if the test suite didn't have to know about vegeta, in case we need to swap it out for a different load testing tool in the future. 

Can the RunLoadTest method accept a generic Target or Request object instead?

kate-osborn commented on tests/README.md:
```diff
@@ -6,22 +6,30 @@ are similar to the existing [conformance tests](../conformance/README.md), but w
+If running the tests on a VM (`make create-vm-and-run-tests` or `make run-tests-on-vm`):
+
+- The [gcloud CLI](https://cloud.google.com/sdk/docs/install)
+- A GKE cluster allowing control plane network access from VMs in the same region
```
Can this also be automated, or can we provide an example of how to set this up?

pleshakov commented on tests/scripts/create-and-setup-gcp-vm.sh:
```diff
@@ -0,0 +1,40 @@
+    --network=default \
+    --action=ALLOW \
+    --rules=tcp:22 \
+    --source-ranges=$(curl -sS -4 icanhazip.com)/32 \
```
I'm a bit concerned that use some not-well-known external servics, especially from security perspective. 
would it be too difficult if we make the source range configurable in the vars file?

pleshakov commented on tests/scripts/vars.env-example:
```diff
@@ -0,0 +1,14 @@
+GKE_SVC_ACCOUNT=<service account with k8s admin permissions>
+IMAGE=projects/debian-cloud/global/images/debian-11-bullseye-v20231115
+FIREWALL_RULE_NAME=<name of the firewall rule that will be created allowing SSH access to the VM>
+TAGS=<network tags>
```
I recommend renaming to NETWORK_TAGS, because TAGS looks almost the same as TAG

pleshakov commented on tests/scripts/vars.env-example:
```diff
@@ -0,0 +1,14 @@
+VM_NAME=ngf-auto-testing
```
+1 to the warning.
but because it doesn't include any credentials, even if it leaks, I don't think it is too bad.

pleshakov commented on tests/Makefile:
```diff
@@ -22,10 +30,26 @@ load-images: ## Load NGF and NGINX images on configured kind cluster
+create-vm-and-run-tests: create-and-setup-vm run-tests-on-vm ## Create and setup a GCP VM for tests and run the tests
+
+.PHONY: cleanup-vm
+cleanup-vm: ## Delete the test GCP VM and delete the firewall rule
```
would it make sense to move this code to a script in scripts folder?
reasoning - this way, logic about names and stuff - all in one place - in the scripts folder

pleshakov commented on tests/framework/ngf.go:
```diff
@@ -110,6 +111,9 @@ func InstallNGF(cfg InstallationConfig, extraArgs ...string) ([]byte, error) {
 	if cfg.ServiceType != "" {
 		args = append(args, formatValueSet("service.type", cfg.ServiceType)...)
+		if cfg.ServiceType == "LoadBalancer" && cfg.IsGKEInternalLB {
+			args = append(args, formatValueSet(`service.annotations.networking\.gke\.io\/load-balancer-type`, "Internal")...)
```
I wonder why escaping is needed before .gke and .io, but not before .annotations or .networking?

ciarams87 replied:
Because `annotations` here corresponds to the Yaml object, but all the remaining dots are part of the annotation itself. The dot before `networking` indicates to helm a new `break`. The formatted result is like this:
```
<...>
kind: Service
metadata:
  <...>
  annotations:
    networking.gke.io/load-balancer-type: "Internal"
<...>
```

pleshakov replied:
thanks for clarification!!

ciarams87 commented on tests/suite/dataplane_perf_test.go:
```diff
@@ -0,0 +1,108 @@
+		Expect(err).To(BeNil())
+	})
+
+	It("writes the system info to a results file", func() {
```
Good point!

kate-osborn reviewed (approved):
🚀 🚀 🚀 

I tested this on GCP and it worked like a charm!
